File: src/tools/connect.ts
import { type ToolDefinition, type ToolResult } from "../core/types.js";

export function getTools(): ToolDefinition[] {
  const make = (n: string, d: string, s: Record<string, unknown>, fn: (p: any, ctx: any) => Promise<ToolResult>): ToolDefinition => ({
    name: n, description: d, inputSchema: s as any, execute: fn,
  });
  return [
    make("connect_status", "Get connection status. Read-only.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("query { connect { status } }") })),
    make("connect_remote_access", "Get remote access info. Read-only.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("query { connect { remoteAccess { enabled url } } }") })),
    make("connect_cloud", "Get cloud status. Read-only.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("query { connect { cloud { status email } } }") })),
    make("connect_sign_in", "Sign in to Connect.", { type:"object", properties:{ email:{type:"string"}, password:{type:"string"} }, required:["email","password"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation SignIn($email: String!, $password: String!) { connect { signIn(email: $email, password: $password) { status } } }", p) })),
    make("connect_sign_out", "Sign out of Connect.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation { connect { signOut { status } } }") })),
    make("connect_devices", "Get paired devices. Read-only.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("query { connect { devices { id name type status } } }") })),
    make("connect_pair_device", "Pair a device.", { type:"object", properties:{ name:{type:"string"} }, required:["name"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation PairDevice($name: String!) { connect { pairDevice(name: $name) { id code } } }", p) })),
    make("connect_remove_device", "Remove a device.", { type:"object", properties:{ device_id:{type:"string"} }, required:["device_id"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation RemoveDevice($id: ID!) { connect { removeDevice(id: $id) { status } } }", p) })),
  ];
}

File: go/internal/tools/connect.go
package connect

import (
	"context"
	"fmt"

	"github.com/hektyc/unraid-mcp-server/go/internal/config"
	"github.com/hektyc/unraid-mcp-server/go/internal/guards"
	"github.com/hektyc/unraid-mcp-server/go/internal/mcp"
)

func RegisterTools(s *mcp.Server, cfg *config.Config) {
	s.RegisterTool(&mcp.ToolDef{
		Name:        "connect_status",
		Description: "Get connection status. Read-only.",
		Query:       `query { connect { status } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "connect_sign_in",
		Description: "Sign in to Connect.",
		Query:       `mutation SignIn($email: String!, $password: String!) { connect { signIn(email: $email, password: $password) { status } } }`,
		Params: map[string]string{
			"email": "string", // required=true
			"password": "string", // required=true
		},
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "connect_sign_out",
		Description: "Sign out of Connect.",
		Query:       `mutation { connect { signOut { status } } }`,
	})
}